### ISSUE TITLE
AGS 4: new compiler does not initialize local vars properly under certain conditions

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1996]
@fernewelten 

I found that the new compiler can miss a stack write operation somehow when initializing a local variable with another local variable. I'm not completely certain about conditions, but one minimal case when such error occurs is this:

```
function game_start()
{
    Character* c;
    Character* c2 = c;
    Display("%p", c2);
}
```

Here the first variable `c` is correctly initialized as 0, but the second is not initialized at all. The SCMD_ADD seem to be called when allocating it, but the actual write to stack does not happen (?). Because of that the `c2` keeps random garbage from the virtual stack.

The reason why this cannot be noticed earlier is because engine was memsetting the stack mem upon SCMD_ADD with zeroes. As in similar cases the variable should be 0 anyway, this would work by coincidence. But as of pr #1990 we did number of speed optimizations to the script VM, and this memset was removed, causing the problem to surface.

I toggled "instruction dump" in the engine, and following was the output:
```
Line   3, IP:       2 (SP:04008824) thisaddr 0
Line   4, IP:       6 (SP:04008824) load.sp.offs 0
Line   4, IP:       8 (SP:04008824) memwrite.ptr.0
Line   4, IP:       9 (SP:04008824) addi sp, 4
Line   5, IP:      14 (SP:04008838) load.sp.offs 4
Line   5, IP:      16 (SP:04008838) memread.ptr ax
Line   5, IP:      18 (SP:04008838) load.sp.offs 0
Line   5, IP:      20 (SP:04008838) memwrite.ptr ax
Line   5, IP:      22 (SP:04008838) addi sp, 4
Line   6, IP:      27 (SP:0400884C) load.sp.offs 4
Line   6, IP:      29 (SP:0400884C) memread.ptr ax
```

It's curious that if we initialize the first variable with any value, then this suddenly works correctly:
```
function game_start()
{
    Character* c = player;
    Character* c2 = c;
    Display("%p", c2);
}
```

Maybe I'm missing something here myself, or this is a mistake in the engine and not the compiler. I will make a break and look into this again.

---

I must clarify further, as one may wonder why would I bother with the seemingly useless code.
With PR #1923 I am adding support for nested managed structs (pointers allowed inside managed structs).
When testing some code, I met above error in the function that looked like this:

```
managed struct ListItem
{
    ListItem* prev;
    ListItem* next;
    
    String name;
    int data;
};

ListItem* newItem(ListItem* add_after, String name, int data)
{
    ListItem* item = new ListItem;
    item.name = name;
    item.data = data;
    if (add_after != null)
    {
        ListItem* next = add_after.next; // <--------- MISTAKE: no actual value is written to stack
        add_after.next = item;
        item.prev = add_after;
        if (next != null) // <------------- ERROR: 'next' contains garbage
        {
            next.prev = item;
            item.next = next;
        }
    }
    return item;
}
```

It appears that `ListItem* next` is never initialized properly in above script, for the same or similar reason.

Here's the "dump" for this function `newItem` run with `add_after` not null:
```
Line   3, IP:       2 (SP:0415D3B8) thisaddr 0
Line   3, IP:       4 (SP:0415D3B8) load.sp.offs 8
Line   3, IP:       6 (SP:0415D3B8) memread4 ax
Line   3, IP:       8 (SP:0415D3B8) meminit.ptr ax
Line   3, IP:      10 (SP:0415D3B8) load.sp.offs 12
Line   3, IP:      12 (SP:0415D3B8) memread4 ax
Line   3, IP:      14 (SP:0415D3B8) meminit.ptr ax
Line   4, IP:      18 (SP:0415D3B8) newuserobject2 ax, 1487, 16
Line   4, IP:      22 (SP:0415D3B8) load.sp.offs 0
Line   4, IP:      24 (SP:0415D3B8) memwrite.ptr ax
Line   4, IP:      26 (SP:0415D3B8) addi sp, 4
Line   5, IP:      31 (SP:0415D3CC) load.sp.offs 16
Line   5, IP:      33 (SP:0415D3CC) memread.ptr ax
Line   5, IP:      35 (SP:0415D3CC) load.sp.offs 4
Line   5, IP:      37 (SP:0415D3CC) memread.ptr mar
Line   5, IP:      39 (SP:0415D3CC) checknull.ptr
Line   5, IP:      40 (SP:0415D3CC) addi mar, 8
Line   5, IP:      43 (SP:0415D3CC) memwrite.ptr ax
Line   6, IP:      47 (SP:0415D3CC) load.sp.offs 20
Line   6, IP:      49 (SP:0415D3CC) memread4 ax
Line   6, IP:      51 (SP:0415D3CC) load.sp.offs 4
Line   6, IP:      53 (SP:0415D3CC) memread.ptr mar
Line   6, IP:      55 (SP:0415D3CC) checknull.ptr
Line   6, IP:      56 (SP:0415D3CC) addi mar, 12
Line   6, IP:      59 (SP:0415D3CC) memwrite4 ax
Line   7, IP:      63 (SP:0415D3CC) load.sp.offs 12
Line   7, IP:      65 (SP:0415D3CC) memread.ptr ax
Line   7, IP:      67 (SP:0415D3CC) push ax
Line   7, IP:      69 (SP:0415D3E0) movl ax, 0
Line   7, IP:      72 (SP:0415D3E0) pop bx
Line   7, IP:      74 (SP:0415D3CC) cmpne bx, ax
Line   7, IP:      77 (SP:0415D3CC) mov bx, ax
Line   7, IP:      80 (SP:0415D3CC) jzi 106
Line   9, IP:      84 (SP:0415D3CC) load.sp.offs 12
Line   9, IP:      86 (SP:0415D3CC) memread.ptr mar
Line   9, IP:      88 (SP:0415D3CC) checknull.ptr
Line   9, IP:      89 (SP:0415D3CC) addi mar, 4
Line   9, IP:      92 (SP:0415D3CC) memread.ptr ax
Line   9, IP:      94 (SP:0415D3CC) load.sp.offs 0
Line   9, IP:      96 (SP:0415D3CC) memwrite.ptr ax
Line   9, IP:      98 (SP:0415D3CC) addi sp, 4
Line  10, IP:     103 (SP:0415D3E0) load.sp.offs 8
Line  10, IP:     105 (SP:0415D3E0) memread.ptr ax
Line  10, IP:     107 (SP:0415D3E0) load.sp.offs 16
Line  10, IP:     109 (SP:0415D3E0) memread.ptr mar
Line  10, IP:     111 (SP:0415D3E0) checknull.ptr
Line  10, IP:     112 (SP:0415D3E0) addi mar, 4
Line  10, IP:     115 (SP:0415D3E0) memwrite.ptr ax
Line  11, IP:     119 (SP:0415D3E0) load.sp.offs 16
Line  11, IP:     121 (SP:0415D3E0) memread.ptr ax
Line  11, IP:     123 (SP:0415D3E0) load.sp.offs 8
Line  11, IP:     125 (SP:0415D3E0) memread.ptr mar
Line  11, IP:     127 (SP:0415D3E0) checknull.ptr
Line  11, IP:     128 (SP:0415D3E0) memwrite.ptr ax
Line  12, IP:     132 (SP:0415D3E0) load.sp.offs 4
Line  12, IP:     134 (SP:0415D3E0) memread.ptr ax
```
here it crashes with "cannot retrieve pointer" error, because `next` variable contains garbage.

**Comment by ivan-mogilko** [comment]
I am investigating this again; not fully certain yet, but i have suspicions that something is wrong in the interpreter rather than in the generated bytecode.

In any case, for the reference, the variant of bytecode for the test case 1 generated by the old compiler:
```
Line   3, IP:       2 (SP:040AEF5C) thisaddr 0
Line   4, IP:       6 (SP:040AEF5C) mov sp, mar
Line   4, IP:       9 (SP:040AEF5C) zeromem 4
Line   4, IP:      11 (SP:040AEF5C) addi sp, 4
Line   5, IP:      16 (SP:040AEF70) load.sp.offs 4
Line   5, IP:      18 (SP:040AEF70) push mar
Line   5, IP:      20 (SP:040AEF84) pop mar
Line   5, IP:      22 (SP:040AEF70) memread.ptr ax
Line   5, IP:      24 (SP:040AEF70) mov sp, mar
Line   5, IP:      27 (SP:040AEF70) meminit.ptr ax
Line   5, IP:      29 (SP:040AEF70) addi sp, 4
Line   6, IP:      34 (SP:040AEF84) load.sp.offs 4
Line   6, IP:      36 (SP:040AEF84) push mar
Line   6, IP:      38 (SP:040AEF98) pop mar
Line   6, IP:      40 (SP:040AEF84) memread.ptr ax  <-------- new compiler's code crashes here
Line   6, IP:      42 (SP:040AEF84) farpush ax
Line   6, IP:      44 (SP:040AEF84) movl ax, 212
Line   6, IP:      47 (SP:040AEF84) farpush ax
Line   6, IP:      49 (SP:040AEF84) setfuncargs 2
Line   6, IP:      51 (SP:040AEF84) movl ax, 468
Line   6, IP:      54 (SP:040AEF84) farcall ax
```

**Comment by ivan-mogilko** [comment]
Alright, I found what's wrong.

The cause of this error is the combination of two things:
1) old optimization in the script interpreter, in the SCMD_MEMWRITEPTR, where it does not assign a dynamic object's pointer to the reg if the handle is matching the one already stored in it.
See: 
https://github.com/adventuregamestudio/ags/blob/cd5b2c0d4d32cd9010a19e907fde8465b0c7090e/Engine/script/cc_instance.cpp#L1160-L1163
2) The interpreter's quirk, where it would consider ADD over SP register (stackpointer) a data allocation, and if the stored ScriptValue is undefined, it would convert it into "array pointer" pointing to the raw data buffer. But as nothing was written to that buffer, and it was not zeroed, this practically would create a variable containing garbage.

So, apparently this may be fixed by removing `registers[SREG_MAR].WriteInt32(newHandle);` statement out of the condition `if (handle != newHandle)` and always reset the reg.

**Comment by ivan-mogilko** [comment]
Fixed by 4f88b4986a331955257af36c517b02476ec2ef23
